File: algorithms/bubble_sort/haskell/bubble_sort.hs
import Debug.Trace

main :: IO()
main = do
  let sorted = bubbleSort [6, 5, 3, 1, 8, 7, 2, 4] :: [Integer]
  print sorted

bubbleSort :: (Ord a, Show a) => [a] -> [a]
--bubbleSort lst | trace ("sorting: " ++ show lst) False = undefined
bubbleSort [] = []
bubbleSort [x] = [x]
bubbleSort (x:y:rest) =
  bubbleSort (init bubbled) ++ [last bubbled]
  where
    (first, second) = if x > y then (y,x) else (x,y)
    bubbled = first : bubbleSort (second:rest)

File: algorithms/bubble_sort/php/bubble_sort.php
<?php

function bubble_sort($arr) {
    $count = count($arr);

    for ($i = 0; $i < $count; $i++) {
        for ($j = $i + 1; $j < $count; $j++) {
            if ($arr[$i] > $arr[$j]) {
                [$arr[$i], $arr[$j]] = [$arr[$j], $arr[$i]];
            }
        }   
    }
    
    return $arr;
}

$init = [4,1,6,8,3,5,2,9,7];

print "Before:" . PHP_EOL;
print_r($init);

$sorted = bubble_sort($init);
print "After:" . PHP_EOL;
print_r($sorted);

